api/src/rpc.c

```diff
@@ -0,0 +1,369 @@
+    fn_group->fn_count = fn_count;
+
+    uvisor_pool_queue_enqueue(fn_group_queue(), slot);
+
+    return fn_group;
+}
+
+/* Return a pointer to the function group specified by the fn_ptr_array, or
+ * NULL if that function group isn't found. */
+static uvisor_rpc_fn_group_t * get_function_group(const TFN_Ptr fn_ptr_array[], size_t fn_count)
+{
+    uvisor_pool_slot_t i;
+
+    /* Find the entry with linear search */
+    for (i = 0; i < fn_group_queue()->pool->num; i++) {
+        /* If found: */
+        if (fn_group_array()[i].fn_ptr_array == fn_ptr_array) {
+            /* Return the target. */
+            return &fn_group_array()[i];
+        }
+    }
+
+    /* The entry for the given function pointer wasn't found. */
+    return NULL;
+}
+
+/* Private structure for function group queries */
+typedef struct query_for_fn_group_context {
+    const TFN_Ptr *fn_ptr_array;
+    size_t fn_count;
+} query_for_fn_group_context_t;
+
+static int query_for_fn_group(uvisor_pool_slot_t slot, void * context)
+{
+    query_for_fn_group_context_t * query_context = context;
+    uvisor_rpc_message_t * msg = &incoming_message_array()[slot];
+    size_t i;
+
+    /* See if the target is for a function we can handle. */
+    for (i = 0; i < query_context->fn_count; ++i) {
+        if ((TFN_Ptr) msg->gateway->target == query_context->fn_ptr_array[i]) {
+            /* Yes, we can handle this function call. */
+            return 1;
+        }
+    }
+
+    /* The message wasn't for a function we can handle. */
+    return 0;
+}
+
+/* Return true if and only if we handled an RPC message */
+static int handle_incoming_rpc(uvisor_rpc_fn_group_t * fn_group, int * box_id_caller) {
+    uvisor_pool_slot_t msg_slot;
+    uvisor_rpc_message_t * msg;
+    query_for_fn_group_context_t context;
+
+    context.fn_ptr_array = fn_group->fn_ptr_array;
+    context.fn_count = fn_group->fn_count;
+
+    msg_slot = uvisor_pool_queue_find_first(incoming_message_todo_queue(), query_for_fn_group, &context);
+    if (msg_slot >= incoming_message_todo_queue()->pool->num) {
+        /* We woke up for no reason. */
+        return 0;
+    }
+
+    /* Dequeue the message */
+    /* TODO Combine finding and dequeing into one operation, so we don't have any
+     * chance to fail dequeing after we find a slot we are interested in. */
+    msg_slot = uvisor_pool_queue_dequeue(incoming_message_todo_queue(), msg_slot);
+    if (msg_slot >= incoming_message_todo_queue()->pool->num)
+    {
+        /* In between finding an RPC we could handle and trying to dequeue it,
+         * somebody else dequeued it. */
+        return 0;
+    }
+
+    msg = &incoming_message_array()[msg_slot];
+
+    /* Save the calling box ID before the RPC target function is called, so that the RPC target function
+     * can read that variable to find out what box called it. */
+    if (box_id_caller) {
+        *box_id_caller = msg->other_box_id;
+    }
+    TFN_Ptr fn = (TFN_Ptr) msg->gateway->target;
```

Comment:
@Patater pure paranoia for detecting bugs rather than hacks / memory corruption: run a sanity check whether the gateway is valid (normal gw check including magic etc.)